frontend/src/metabase/lib/utils.ts

```diff
@@ -0,0 +1,216 @@
+    t`one`,
+    t`two`,
+    t`three`,
+    t`four`,
+    t`five`,
+    t`six`,
+    t`seven`,
+    t`eight`,
+    t`nine`,
+  ];
+
+  if (num >= 0 && num <= 9) {
+    return names[num];
+  } else {
+    return "" + num;
+  }
+}
+
+export function uuid() {
+  return (
+    s4() +
+    s4() +
+    "-" +
+    s4() +
+    "-" +
+    s4() +
+    "-" +
+    s4() +
+    "-" +
+    s4() +
+    s4() +
+    s4()
+  );
+}
+
+export function isUUID(uuid: unknown) {
+  return (
+    typeof uuid === "string" &&
+    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/.test(uuid)
+  );
+}
+
+export function isBase64(string: unknown) {
+  return (
+    typeof string === "string" &&
+    /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(
+      string,
+    )
+  );
+}
+
+export function isJWT(string: unknown) {
+  return (
+    typeof string === "string" &&
+    /^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$/.test(string)
+  );
+}
+
+export function isEmail(email: string | undefined | null) {
+  if (email === null || email === undefined) {
+    return false;
+  }
+  return EMAIL_REGEX.test(email);
+}
+
+export function getEmailDomain(email: string) {
+  const match = EMAIL_REGEX.exec(email);
+  return match && match[5];
+}
+
+export function equals(a: unknown, b: unknown) {
+  return _.isEqual(a, b);
+}
+
+export function propertiesEqual(
+  a: Record<string, unknown>,
+  b: Record<string, unknown>,
+  properties = [...Object.keys(a), ...Object.keys(b)],
+) {
+  for (const property of properties) {
+    if (a[property] !== b[property]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function copy(a: unknown) {
+  // FIXME: ugghhhhhhhhh
+  return JSON.parse(JSON.stringify(a));
+}
+
+/**
+ * Converts a metabase version to a list of numeric components, it converts pre-release
+ * components to numbers and pads the numeric part to 4 numbers to make comparison easier
+ */
+export function versionToNumericComponents(version: string): number[] | null {
+  const SPECIAL_COMPONENTS: Record<string, number> = {
+    snapshot: -4,
+    alpha: -3,
+    beta: -2,
+    rc: -1,
+  };
+
+  const regex =
+    /v?(?<ossOrEE>\d+)\.?(?<major>\d+)?\.?(?<minor>\d+)?\.?(?<patch>\d+)?-?(?<label>\D+)?(?<build>\d+)?/;
+
+  const result = regex.exec(version);
+
+  if (!result || !result.groups) {
+    return null;
+  }
+
+  const {
+    ossOrEE,
+    major = 0,
+    minor = 0,
+    patch = 0,
+    label,
+    build = 0,
+  } = result.groups;
+
+  return [
+    ossOrEE,
+    major,
+    minor,
+    patch,
+    SPECIAL_COMPONENTS[label?.toLowerCase()] ?? 0,
+    build,
+  ].map(part => (typeof part === "string" ? parseInt(part, 10) : part));
+}
+
+/**
+ * this should correctly compare all version formats Metabase uses, e.g.
+ * 0.0.9, 0.0.10-snapshot, 0.0.10-alpha1, 0.0.10-rc1, 0.0.10-rc2, 0.0.10-rc10
+ * 0.0.10, 0.1.0, 0.2.0, 0.10.0, 1.1.0
+ */
+export function compareVersions(aVersion: string, bVersion: string): -1 | 0 | 1;
```

Comment:
The type overload is to make our life easier, if we know what we pass can't be undefined then we'll always get a number out of it, handy for sorting (which complains with null returns)